pkg/controllers/user/pipeline/controller/pipelineexecution/logsyncer.go

```diff
@@ -50,54 +50,71 @@ func (s *ExecutionLogSyncer) syncLogs() {
-			e.Labels["pipeline.management.cattle.io/finish"] = "true"
-			if _, err := s.pipelineExecutionLogs.Update(e); err != nil {
+			log.Spec.Message += fmt.Sprintf("\nInvalid pipeline execution name - %s", log.Spec.PipelineExecutionName)
+			if err := s.finishExecutionLog(log); err != nil {
 				logrus.Errorf("Error update pipeline execution log - %v", err)
 				return
 			}
 			continue
 		}
-		execution, err := s.pipelineExecutionLister.Get(e.Namespace, parts[1])
+		execution, err := s.pipelineExecutionLister.Get(log.Namespace, parts[1])
 		if err != nil {
 			logrus.Errorf("Error get pipeline execution - %v", err)
-			e.Spec.Message += fmt.Sprintf("\nError get pipeline execution - %v", err)
-			e.Labels["pipeline.management.cattle.io/finish"] = "true"
-			if _, err := s.pipelineExecutionLogs.Update(e); err != nil {
+			log.Spec.Message += fmt.Sprintf("\nError get pipeline execution - %v", err)
+			if err := s.finishExecutionLog(log); err != nil {
 				logrus.Errorf("Error update pipeline execution log - %v", err)
 				return
 			}
 			continue
 		}
 		//get log if the step started
-		if execution.Status.Stages[e.Spec.Stage].Steps[e.Spec.Step].State == utils.StateWaiting {
+		if execution.Status.Stages[log.Spec.Stage].Steps[log.Spec.Step].State == utils.StateWaiting {
 			continue
 		}
-		logText, err := s.pipelineEngine.GetStepLog(execution, e.Spec.Stage, e.Spec.Step)
+		logText, err := s.pipelineEngine.GetStepLog(execution, log.Spec.Stage, log.Spec.Step)
 		if err != nil {
 			logrus.Errorf("Error get pipeline execution log - %v", err)
-			e.Spec.Message += fmt.Sprintf("\nError get pipeline execution log - %v", err)
-			e.Labels["pipeline.management.cattle.io/finish"] = "true"
-			if _, err := s.pipelineExecutionLogs.Update(e); err != nil {
+			log.Spec.Message += fmt.Sprintf("\nError get pipeline execution log - %v", err)
+			if err := s.finishExecutionLog(log); err != nil {
 				logrus.Errorf("Error update pipeline execution log - %v", err)
+				return
 			}
 			continue
 		}
 
-		e.Spec.Message = logText
-		stepState := execution.Status.Stages[e.Spec.Stage].Steps[e.Spec.Step].State
+		stepState := execution.Status.Stages[log.Spec.Stage].Steps[log.Spec.Step].State
 		if stepState != utils.StateWaiting && stepState != utils.StateBuilding {
-			e.Labels["pipeline.management.cattle.io/finish"] = "true"
+			log.Labels["pipeline.management.cattle.io/finish"] = "true"
 		}
-		if _, err := s.pipelineExecutionLogs.Update(e); err != nil {
+
+		if log.Spec.Message == logText {
+			//only do update on changes
+			continue
+		}
+		log.Spec.Message = logText
+		if _, err := s.pipelineExecutionLogs.Update(log); err != nil {
 			logrus.Errorf("Error update pipeline execution log - %v", err)
 			return
 		}
 	}
 	logrus.Debugf("Sync pipeline execution log complete")
 }
+
+func (s *ExecutionLogSyncer) finishExecutionLog(log *v3.PipelineExecutionLog) error {
+	log.Labels["pipeline.management.cattle.io/finish"] = "true"
```

Comment:
we should implement the constant for this label. I see it being repeated multiple times in existing code